File: ruby/problem_4.rb
#!/bin/env ruby
#
# Problem 4
#
# Author::    mattt
#
require 'ruby/euler_lib'

timed_run{
  num1 = 999
  num2 = 999

  largest_palindrome = 0
  num1.downto(0) do |n|
    num2.downto(0) do |m|
      product = n*m
      if product.palindrome? and product > largest_palindrome
        largest_palindrome = product
      end
    end
    num2 -= 1
  end

  #906609
  puts largest_palindrome
}

File: golang/problem_4.go
package main

func problem4() int {
	num1 := 999
	num2 := 999

	var product int
	var largestPalindrome int

	for n := num1; n >= 0; n-- {
		for m := num2; m >= 0; m-- {
			product = n * m
			if palindrome(product) && product > largestPalindrome {
				largestPalindrome = product
			}
		}
		num2 -= 1
	}
	return largestPalindrome
}
